aggregator/aggregator_test.go

```diff
@@ -10,73 +10,97 @@ import (
 func TestScanner(t *testing.T) {
 	cases := []struct {
 		in    []float64
+		ts    []uint32
 		avg   float64
 		delta float64
 		last  float64
 		max   float64
 		min   float64
 		stdev float64
 		sum   float64
+		deriv float64
 	}{
 		{
 			[]float64{5, 4, 7, 4, 2, 5, 4, 9},
+			[]uint32{1, 2, 3, 4, 5, 6, 7, 8},
 			5,
 			7,
 			9,
 			9,
 			2,
 			2,
 			40,
+			float64(4) / float64(7),
 		},
 		{
 			[]float64{6, 2, 3, 1},
+			[]uint32{1, 2, 3, 4},
 			3,
 			5,
 			1,
 			6,
 			1,
 			1.8708286933869707,
 			12,
+			float64(-5) / float64(3),
+		},
+		// test out of order. this is the same dataset as the first one, but a bit shuffled
+		{
+			[]float64{7, 4, 5, 4, 9, 5, 4, 2},
+			[]uint32{3, 2, 1, 4, 8, 6, 7, 5},
+			5,
+			7,
+			2, // last is the last received one, not the one with last timestamp. we could handle that better perhaps
+			9,
+			2,
+			2,
+			40,
+			float64(4) / float64(7),
 		},
 	}
-	for i, e := range cases {
-		var actual float64
-
-		actual = Avg(e.in)
-		if actual != e.avg {
-			t.Fatalf("case %d AVG - expected %v, actual %v", i, e.avg, actual)
+	testCase := func(i int, name string, in []float64, ts []uint32, exp float64) {
+		procConstr, err := GetProcessorConstructor(name)
+		if err != nil {
+			t.Fatalf("got err %q", err)
 		}
-
-		actual = Delta(e.in)
-		if actual != e.delta {
-			t.Fatalf("case %d DELTA - expected %v, actual %v", i, e.delta, actual)
+		p := procConstr(in[0], ts[0])
+		for i, v := range in[1:] {
+			p.Add(v, ts[i+1])
 		}
-
-		actual = Last(e.in)
-		if actual != e.last {
-			t.Fatalf("case %d LAST - expected %v, actual %v", i, e.last, actual)
+		got, ok := p.Flush()
+		if !ok {
+			t.Fatalf("case %d %s - expected valid output, got null", i, name)
 		}
-
-		actual = Max(e.in)
-		if actual != e.max {
-			t.Fatalf("case %d MAX - expected %v, actual %v", i, e.max, actual)
+		if got != exp {
+			t.Fatalf("case %d %s - expected %v, actual %v", i, name, exp, got)
 		}
+	}
+	for i, e := range cases {
+		testCase(i, "avg", e.in, e.ts, e.avg)
+		testCase(i, "delta", e.in, e.ts, e.delta)
+		testCase(i, "last", e.in, e.ts, e.last)
+		testCase(i, "max", e.in, e.ts, e.max)
+		testCase(i, "min", e.in, e.ts, e.min)
+		testCase(i, "stdev", e.in, e.ts, e.stdev)
+		testCase(i, "sum", e.in, e.ts, e.sum)
+		testCase(i, "derive", e.in, e.ts, e.deriv)
+	}
+}
 
-		actual = Min(e.in)
-		if actual != e.min {
-			t.Fatalf("case %d MIN - expected %v, actual %v", i, e.min, actual)
-		}
+var r float64
```

Comment:
What's this variable for? Don't see it used anywhere other than being assigned in the `BenchmarkProcessorMax` function.

Reply:
it's a safeguard against the compiler just eliminating all the aggregation calls.
code that doesn't have side effects and just returns an output value (which then is not assigned) can just be taken out by the compiler. see the last section at https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go

in practice I haven't seen this happen much, and this behavior may change with different Go versions, so just wanted to be safe.

Reply:
Might be a good idea to document this obscure piece of code :)